### ISSUE TITLE
inference on converted model important cfg parameters

### ISSUE THREAD

**Issue by senemilk** [AlexeyAB/darknet#3092]
hi everyone,

I was working on the conversion of a model from pytorch but I am having an error of 0.26 % compared to initial network.
I think that i might be missing a detail in my cfg file.

here are my cfg file and my config from pytorch:
```
[net]
batch=1
subdivisions=1
height=127
width=127
channels=3
momentum=1.0
epsilon=1e-05

learning_rate = 0.45

[convolutional]
filters=96
size=11
stride=2
padding=0
groups=1
batch_normalize=1
activation = relu

[maxpool]
size=3
stride=2
padding=0

[convolutional]
filters=256
size=5
stride=1
padding=0
groups=1
batch_normalize=1
activation=relu

[maxpool]
size=3
stride=2
padding=0

[convolutional]
filters=384
size=3
stride=1
padding=0
groups=1
batch_normalize=1
activation=relu

[convolutional]
filters=384
size=3
stride=1
padding=0
groups=1
batch_normalize=1
activation=relu

[convolutional]
filters=256
size=3
stride=1
padding=0
groups=1
batch_normalize=1
activation=linear

[convolutional]
filters=2560
size=3
stride=1
padding=0
groups=1
activation=linear
```

and in pytorch : 
```python
class cls1(nn.Module):
    def __init__(self, size=1, feature_out=256, anchor=5):
        configs = [3, 96, 256, 384, 384, 256]
        configs = list(map(lambda x: 3 if x==3 else x*size, configs))
        feat_in = configs[-1]
        super(cls1, self).__init__()
        self.featureExtract = nn.Sequential(
            nn.Conv2d(configs[0], configs[1] , kernel_size=11, stride=2),
            nn.BatchNorm2d(configs[1]),
            nn.MaxPool2d(kernel_size=3, stride=2),
            nn.ReLU(inplace=True),
            nn.Conv2d(configs[1], configs[2], kernel_size=5),
            nn.BatchNorm2d(configs[2]),
            nn.MaxPool2d(kernel_size=3, stride=2),
            nn.ReLU(inplace=True),
            nn.Conv2d(configs[2], configs[3], kernel_size=3),
            nn.BatchNorm2d(configs[3]),
            nn.ReLU(inplace=True),
            nn.Conv2d(configs[3], configs[4], kernel_size=3),
            nn.BatchNorm2d(configs[4]),
            nn.ReLU(inplace=True),
            nn.Conv2d(configs[4], configs[5], kernel_size=3),
            nn.BatchNorm2d(configs[5]),
            )
        
        self.anchor = anchor
        self.feature_out = feature_out
        self.conv_cls1 = nn.Conv2d(feat_in, feature_out*2*anchor, 3)
        self.cfg = {}

    def forward(self, x):
        x_f = self.featureExtract(x)
        return self.conv_cls1(x_f)
```

thanks in advance for you help.

**Comment by AlexeyAB** [comment]
@senemilk Hi,

> I was working on the conversion of a model from pytorch but I am having an error of 0.26 % compared to initial network.

Do you mean that you `Darknet_top5_error - Pytorch_top5_error = 0.26%` or something else?

Also you should know that Darknet uses `pad_mode=SAME` (in TensorFlow terms): https://github.com/pytorch/pytorch/issues/3867

**Comment by senemilk** [comment]
Hi @AlexeyAB ,
 
I mean i used a blank image (array of 255.0) in both networks as input and compared the results.
And the error is 0,26%.

Also there is no padding in the initial network (valid in tensorflow if i am not mistaken) so i added padding=0 in each layer of my cfg file.

**Comment by AlexeyAB** [comment]
* Try to compile Darknet with CUDNN_HALF=0 in the Makefile.

* If it doesn't help, then try to change this line: https://github.com/AlexeyAB/darknet/blob/cce34712f6928495f1fbc5d69332162fc23491b9/src/maxpool_layer_kernels.cu#L88
to `#ifdef CUDNN` and recompile

---

> But there is the dilatation parameter that i dont know how it works on darknet i think it might be the problem

Do you mean you use dilation>1?
Can you show in your code, where is you set dilation parameter?

https://pytorch.org/docs/stable/nn.html

> torch.nn.Conv2d(in_channels, out_channels, kernel_size, stride=1, padding=0, dilation=1, groups=1, bias=True, padding_mode='zeros')

**Comment by senemilk** [comment]
Ok thanks i will check that monday at work and let you know about the results.

No it’s dilatation = 1  (in the nn.conv2d parameters ) but i don’t know if i should write something in the cfg file or if it is 1 by default

**Comment by senemilk** [comment]
hi @AlexeyAB ,

I have another question...

Is there a difference between pytorch's nn.batchnorm2d operation and the forward_batchnorm_layer in darknet ?

because I compared the two (with blank image) and got :

conv2d :
    max Error: 0.0005553961
    pourcentage Error: 0.0001281259317249553  %
    avg erreur :  5.9593024067091985e-05

conv2d + batchnorm2d :
    max Error: 3.941357e-06
    pourcentage Error: 0.0011732527472079024 %
    avg erreur :  9.19663518264834e-07

the percentage is multiplied ten times between the first conv and bn


conv+batchnorm+maxpool+relu + conv :
    max Error: 3.3080578e-06
    pourcentage Error: 0.00036253456734438856 %
    avg erreur :  6.88468012958765e-07

conv+batchnorm+maxpool+relu + conv+batchnorm :
    max Error: 0.0001707077
    pourcentage Error: 0.05651725443756156 %
    avg erreur :  5.143832564353943e-05

and around 150 times between the second conv and bn layers

**Comment by AlexeyAB** [comment]
> No it’s dilatation = 1 (in the nn.conv2d parameters ) but i don’t know if i should write something in the cfg file or if it is 1 by default

Darknet uses `dilatation = 1` automatically.

----

Darknet's batch-norm does: https://machinethink.net/blog/object-detection-with-yolo/
```
        gamma * (out[j] - mean)
bn[j] = ---------------------- + beta
            sqrt(variance)
```

Do you get output of layers by using Darknet on CPU (GPU=0 AVX=0 OPENMP=0) or GPU (GPU=1 CUDNN=1)?

**Comment by senemilk** [comment]
I am using darknet on CPU (GPU=0 AVX=0 OPENMP=0)  and i tested GPU = 1 CUDNN = 0 and it's just faster but the result doesn't change

**Comment by AlexeyAB** [comment]
@senemilk 

* Do you try to use this Yolo implementation on PyTorch? https://github.com/ultralytics/yolov3

* Check one more time that you set `AVX=0 OPENMP=0` in the Makefile

* if you use `./darknet detector test ...` then try to comment this line: https://github.com/AlexeyAB/darknet/blob/cce34712f6928495f1fbc5d69332162fc23491b9/src/detector.c#L1245
It removes batch-normalization layer by fusing `conv+batch_norm --to--> conv` https://github.com/AlexeyAB/darknet/blob/cce34712f6928495f1fbc5d69332162fc23491b9/src/network.c#L957-L993

* after that, the batch-norm will do this
    * firstly https://github.com/AlexeyAB/darknet/blob/cce34712f6928495f1fbc5d69332162fc23491b9/src/blas.c#L129-L140
    * then https://github.com/AlexeyAB/darknet/blob/cce34712f6928495f1fbc5d69332162fc23491b9/src/batchnorm_layer.c#L150

----

* Check that you correctly load the same weights in both Darknet and PyTorch: https://github.com/AlexeyAB/darknet/blob/cce34712f6928495f1fbc5d69332162fc23491b9/src/parser.c#L990-L996

**Comment by senemilk** [comment]
hi @AlexeyAB 
I am not using https://github.com/ultralytics/yolov3 , just the network I described in my initial post.

I checked again for AVX and OPENMP
Also I am using network_predict instead of  ./darknet detector  so it is already performing normalize_cpu then scale_bias

I did a comparison of my weights and they are the same value.
Just that I used numpy.float32 type when storing them to my weightfile  

I checked and noticed that darknet was using single precision for its operations but the weights in pytorch are actually in python's float format which is equal to the double type in C.

So I think that I need to change my darknet so that it can perform double precision operations but the problem is the solution I find require to change almost all of darknet's files

Do you think I can do this without having to change the whole darknet ?

Thank you

**Comment by AlexeyAB** [comment]
> Do you think I can do this without having to change the whole darknet ?

No.

I think it is easier to change your Python implementation, since PyTorch uses single-precision, especially on GPU (CUDA and cuDNN).


Also if you are interested in only forward-inference (without training) and only on CPU (withoug GPU), then you can use https://github.com/AlexeyAB/yolo2_light and change just this one file: https://github.com/AlexeyAB/yolo2_light/blob/master/src/yolov2_forward_network.c

**Comment by senemilk** [comment]
@AlexeyAB 

yes It turn out single precision was enough.
Actually the results I got were already good, i had a problem in my inference code...
I corrected it and it is now detecting normally.

thank you for your help, i appreciate it.
